### PR TITLE
add support to configure max latency

### PR DESCRIPTION
**Pull request by cctdaniel**
add configurable max latency in slots between send and receive

for price feeds with more publishers this can be reduced instead of using the default 25 value to reduce latency

### CODE REVIEW

File: program/c/src/oracle/oracle.h [hunk 1]
```diff
@@ -195,8 +195,9 @@ typedef struct pc_price
   pc_ema_t        twac_;              // time-weighted average conf interval
   int64_t         timestamp_;         // unix timestamp of aggregate price
   uint8_t         min_pub_;           // min publishers for valid price
-  int8_t          drv2_;              // space for future derived values
-  int16_t         drv3_;              // space for future derived values
+  int8_t          message_sent_;      // flag to indicate if the current aggregate price has been sent as a message to the message buffer, 0 if not sent, 1 if sent
```
Comment (cctdaniel) [inline program/c/src/oracle/oracle.h:198]: drive-by: this was already changed in the rust program previously but wasn't changed here

File: program/c/src/oracle/oracle.h [hunk 2]
```diff
@@ -195,8 +195,9 @@ typedef struct pc_price
...
+  uint8_t         max_latency_;       // configurable max latency in slots between send and receive
```
Comment (cctdaniel) [inline program/c/src/oracle/oracle.h:199]: we only need `uint8_t` for `max_latency_` so splitting it to `uint8_t` and `int8_t`

File: program/rust/src/tests/test_twap.rs [hunk 3]
```diff
@@ -19,20 +19,24 @@ use {
...
         DataEvent {
-            slot_gap: u64::from(u8::arbitrary(g)) + 1, /* Slot gap is always > 1, because there
-                                                        * has been a succesful aggregation */
-            price:    i64::arbitrary(g),
-            conf:     u64::arbitrary(g),
+            slot_gap:    u64::from(u8::arbitrary(g)) + 1, /* Slot gap is always > 1, because there
+                                                           * has been a succesful aggregation */
+            price:       i64::arbitrary(g),
+            conf:        u64::arbitrary(g),
+            max_latency: 0,
```
Comment (guibescos) [inline program/rust/src/tests/test_twap.rs:39]: I think here you want something abritary to extend the test coverage

File: program/rust/src/tests/test_twap.rs [hunk 4]
```diff
@@ -87,12 +96,18 @@ impl PriceCumulative {
...
     pub fn check_num_down_slots(&self, data: &[DataEvent]) {
         assert_eq!(
-            data.iter()
-                .fold(0, |acc, x| if x.slot_gap > PC_MAX_SEND_LATENCY.into() {
-                    acc + (x.slot_gap - PC_MAX_SEND_LATENCY as u64)
+            data.iter().fold(0, |acc, x| {
+                let latency_threshold = if x.max_latency == 0 {
+                    PC_MAX_SEND_LATENCY.into()
+                } else {
+                    x.max_latency as u64
```
Comment (guibescos) [inline program/rust/src/tests/test_twap.rs:103]: `.into()` seems cleaner, also I don't think this pathway is ever explored because DataEvent has max_latency equal to 0 always

File: program/rust/src/tests/test_upd_aggregate.rs [hunk 5]
```diff
@@ -269,6 +286,133 @@ fn test_upd_aggregate() {
         assert_eq!(price_data.prev_conf_, 85);
         assert_eq!(price_data.prev_timestamp_, 5);
     }
+
+    // check max_latency_ = 0 defaults to PC_MAX_SEND_LATENCY
```
Comment (guibescos) [inline program/rust/src/tests/test_upd_aggregate.rs:290]: mm arguably this is already checked above?

File: program/c/src/oracle/upd_aggregate.h [hunk 6]
```diff
@@ -175,7 +175,7 @@ static inline bool upd_aggregate( pc_price_t *ptr, uint64_t slot, int64_t timest
            // No overflow for INT64_MIN+conf or INT64_MAX-conf as 0 < conf < INT64_MAX
            // These checks ensure that price - conf and price + conf do not overflow.
            (int64_t)0 < conf && (INT64_MIN + conf) <= price && price <= (INT64_MAX-conf) &&
-           slot_diff >= 0 && slot_diff <= PC_MAX_SEND_LATENCY ) {
```
Comment (guibescos) [inline program/c/src/oracle/upd_aggregate.h:178]: Was it voluntary to remove` slot_diff >= 0`

File: program/rust/src/tests/test_upd_aggregate.rs [hunk 7]
```diff
@@ -269,6 +286,133 @@ fn test_upd_aggregate() {
...
+        assert_eq!(price_data.prev_price_, 245);
+        assert_eq!(price_data.prev_conf_, 85);
+        assert_eq!(price_data.prev_timestamp_, 5);
+    }
+
+    // ensure the update occurs within the PC_MAX_SEND_LATENCY limit of 25 slots, allowing the aggregated price to reflect both p5 and p6 contributions
+    {
+        let mut price_data = load_checked::<PriceAccount>(&price_account, PC_VERSION).unwrap();
+        price_data.num_ = 2;
+        price_data.last_slot_ = 1025;
```
Comment (guibescos) [inline program/rust/src/tests/test_upd_aggregate.rs:326]: I think this would be more readable if you don't update `last_slot` and `agg_.pub_slot` and you keep them equal to 1000 like above.
To create situations where one of the prices expires you just need to tune the `pub_slot_` parameter of the price component.
Here you're modifying two parameters at a time so it's a bit confusing

File: program/c/src/oracle/upd_aggregate.h (hunk 6 above)
Comment (ali-bahjati) [reply program/c/src/oracle/upd_aggregate.h:178]: I think we need to keep it to encourage a publisher sending a price in future slot and get that price included for all slots.

File: program/rust/src/tests/test_twap.rs [hunk 8]
```diff
@@ -163,7 +198,12 @@ fn test_twap_unit() {
     );
     assert_eq!(
         price_cumulative_overflow.num_down_slots,
-        u64::MAX - PC_MAX_SEND_LATENCY as u64
+        u64::MAX
+            - if TEST_MAX_LATENCY == 0 {
```
[the remark this answers is not in the document]
Comment (ali-bahjati) [reply program/rust/src/tests/test_twap.rs:202]: generally I recommend adding different test cases measuring different values for max_latency in TWAP including 0.

**Review by ali-bahjati — COMMENTED** [review]
Very nice! I will look again after you addressed the feedbacks (just to be extra careful).

File: program/c/src/oracle/upd_aggregate.h (hunk 6 above)
Comment (cctdaniel) [reply program/c/src/oracle/upd_aggregate.h:178]: this is redundant, there is a check slightly above to return false if `agg_.pub_slot_` is greater than current slot
```
  // only re-compute aggregate in next slot
  if ( slot <= ptr->agg_.pub_slot_ ) {
    return false;
  }
```

Comment (guibescos) [reply program/c/src/oracle/upd_aggregate.h:178]: ok

File: program/rust/src/tests/test_upd_aggregate.rs [hunk 9]
```diff
@@ -269,6 +279,134 @@ fn test_upd_aggregate() {
...
+        ));
+    }
+
+    {
+        let price_data = load_checked::<PriceAccount>(&price_account, PC_VERSION).unwrap();
+
+        assert_eq!(price_data.max_latency_, 0);
+        assert_eq!(price_data.agg_.price_, 445);
+        assert_eq!(price_data.agg_.conf_, 55);
+        assert_eq!(price_data.twap_.val_, 168);
```
Comment (guibescos) [inline program/rust/src/tests/test_upd_aggregate.rs:339]: I'm a bit confused about these twaps, imo something that can't be checked manually easily shouldn't be here.

**Review by guibescos — DISMISSED** [review]

File: program/rust/src/tests/test_upd_aggregate.rs (hunk 9 above)
Comment (cctdaniel) [reply program/rust/src/tests/test_upd_aggregate.rs:339]: are you suggesting to remove the twap and twac checks here?

**Review by guibescos — DISMISSED** [review]

File: program/c/src/oracle/upd_aggregate.h [hunk 10]
```diff
@@ -175,7 +175,7 @@ static inline bool upd_aggregate( pc_price_t *ptr, uint64_t slot, int64_t timest
...
+           slot_diff <= (ptr->max_latency_ ? ptr->max_latency_ : PC_MAX_SEND_LATENCY) ) {
```
Comment (ali-bahjati) [inline program/c/src/oracle/upd_aggregate.h:178]: I recommend setting the max latency as a variable outside this if (just for more code clarity)

**Review by ali-bahjati — APPROVED** [review]
Nice! I left some minor comments.

**Review by ali-bahjati — DISMISSED** [review]
Nice! I left some minor comments.

**Review by ali-bahjati — APPROVED** [review]
